hydroflow/tests/surface_async.rs

```diff
@@ -258,8 +255,113 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
         _ = tokio::time::sleep(Duration::from_secs(1)) => (),
     };
 
-    let seen: Vec<_> = hydroflow::util::collect_ready(seen_recv).await;
+    let seen: Vec<_> = collect_ready(seen_recv);
     rassert_eq!(&std::array::from_fn::<_, MAX, _>(|i| i), &*seen)?;
 
     Ok(())
 }
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_sink_bounded_channel() {
+    // In this example we use an _bounded_ channel for our `Sink`. This is for demonstration only,
+    // instead you should use [`hydroflow::util::unbounded_channel`]. A bounded channel results in
+    // `Hydroflow` buffering items internally instead of within the channel.
+    let (send, recv) = tokio::sync::mpsc::channel::<usize>(5);
+    let send = tokio_util::sync::PollSender::new(send);
+    let mut recv = tokio_stream::wrappers::ReceiverStream::new(recv);
+
+    let mut flow = hydroflow_syntax! {
+        source_iter(0..10) -> dest_sink(send);
+    };
+    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
+        .await
+        .expect_err("Expected time out");
+
+    // Only 5 elemts received due to buffer size
+    let out: Vec<_> = collect_ready(&mut recv);
+    assert_eq!(&[0, 1, 2, 3, 4], &*out);
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_sink_duplex() {
+    use bytes::Bytes;
+    use tokio::io::AsyncReadExt;
+    use tokio_util::codec;
+
+    // Like a channel, but for a stream of bytes instead of discrete objects.
+    let (asyncwrite, mut asyncread) = tokio::io::duplex(256);
+    // Now instead handle discrete byte lists by length-encoding them.
+    let mut sink = codec::LengthDelimitedCodec::builder()
+        // Use 1 byte len field (max 255) so we don't have to worry about endianness.
+        .length_field_length(1)
+        .new_write(asyncwrite);
+
+    let mut flow = hydroflow_syntax! {
+        source_iter([
+            Bytes::from_static(b"hello"),
+            Bytes::from_static(b"world"),
+        ]) -> dest_sink(&mut sink);
+    };
+    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
+        .await
+        .expect_err("Expected time out");
+
+    let mut buf = Vec::<u8>::new();
+    asyncread.read_buf(&mut buf).await.unwrap();
+    // `\x05` is length prefix of "5".
+    assert_eq!(b"\x05hello\x05world", &*buf);
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_asyncwrite_duplex() {
```

Comment:
similarly complicated. I get uncomfortable when I see calls to raw `tokio::io` and codecs, to be honest.